### ISSUE TITLE
Training speed issue

### ISSUE THREAD

**Issue by ybrrraway** [AntResearchNLP/ViLaSR#7]
I adopted dual-node training, but observed that the speed was the same as that of single-node training. I modified some script parameters. I don't know if it has any impact.

------------------------------------------------------------------
My script is like this：
`#!/bin/bash
set -x

MODEL_PATH=checkpoints/reflective/qwen2_5_vl-7b_vilasr_reflective              # checkpoint after reflective rejection sampling

export PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True
python3 -m verl.trainer.main \
    config=./train/rl/config_grpo_multi.yaml \
    data.train_files=/home/james159/hyb/ViLaSR/ViLaSR-data/rl/vilasr_rl_data_abs.jsonl \
    data.val_files=/home/james159/hyb/ViLaSR/ViLaSR-data/rl/vilasr_rl_data_abs.jsonl \
    data.max_response_length=13312 \
    algorithm.kl_coef=0.0 \
    algorithm.disable_kl=true \
    algorithm.use_kl_loss=false \
    worker.actor.model.model_path=${MODEL_PATH} \
    worker.rollout.tensor_parallel_size=1 \
    worker.rollout.enable_chunked_prefill=false \
    worker.rollout.gpu_memory_utilization=0.35 \
    worker.rollout.n=8 \
    worker.rollout.num_llm_calls_available=10 \
    worker.rollout.limit_images=45 \
    trainer.experiment_name=stage3 \
    trainer.n_gpus_per_node=8 \
    trainer.nnodes=2 \
    trainer.save_freq=50 \
    trainer.save_checkpoint_path='checkpoints/rl_board/qwen2_5_vl-7b_vilasr_grpo' \`

------------------------------------------------------------------

------------------------------------------------------------------
the config_grpo_multi.yaml:
`data:
  train_files: hiyouga/math12k@train
  val_files: hiyouga/math12k@test
  prompt_key: question
  answer_key: answer
  image_key: image_path
  max_prompt_length: 5120
  max_response_length: 16384            # last: 16384, 8192
  rollout_batch_size: 32                  #  origin: 512
  val_batch_size: 16                       # -1, batch_size=len(self.val_dataset) if self.config.data.val_batch_size == -1 else self.config.data.val_batch_size,
  shuffle: true
  seed: 1
  max_pixels: 200704                  # 256*28*28=448*448=200704; 512*28*28 = 633*633; origin: 4194304
  min_pixels: 16384                        # 262144

  
algorithm:
  adv_estimator: grpo
  disable_kl: false     # apply kl to reward
  use_kl_loss: true     # apply kl to loss
  kl_penalty: low_var_kl
  kl_coef: 1.0e-2

worker:
  actor:
    global_batch_size: 16                                    # 128
    micro_batch_size_per_device_for_update: 1                  # origin: 4; 4 or 8 or 16
    micro_batch_size_per_device_for_experience: 1              #  16
    max_grad_norm: 1.0
    padding_free: true
    ulysses_sequence_parallel_size: 1
    model:
      model_path: checkpoints/reflective/qwen2_5_vl-7b_vilasr_reflective 
      enable_gradient_checkpointing: true
      trust_remote_code: false
      freeze_vision_tower: false
    optim:
      lr: 1.0e-6
      weight_decay: 1.0e-2
      strategy: adamw  # {adamw, adamw_bf16}
      lr_warmup_ratio: 0.0
    fsdp:
      enable_full_shard: true
      enable_cpu_offload: false
      enable_rank0_init: true
    offload:
      offload_params: false  # true: more CPU memory; false: more GPU memory
      offload_optimizer: false  # true: more CPU memory; false: more GPU memory

  rollout:
    temperature: 1.0
    n: 8                                # origin: 5
    gpu_memory_utilization: 0.60      # origin: 0.60
    enforce_eager: false
    enable_chunked_prefill: false
    tensor_parallel_size: 1
    limit_images: 0
    val_override_config:
      temperature: 0.5
      n: 1
    max_num_batched_tokens: 32768       # > model_len last 20480, 16384
    # prompt_length
    # response_length
    single_turn_response_length: 1024

  ref:
    fsdp:
      enable_full_shard: true
      enable_cpu_offload: false  # true: more CPU memory; false: more GPU memory
      enable_rank0_init: true
    offload:
      offload_params: false

  reward:
    reward_type: function
    compute_score: gcot # math

trainer:
  total_episodes: 1                             # origin 15
  logger: ["console", "tensorboard"]
  project_name: easy_r1
  experiment_name: easy_r1
  n_gpus_per_node: 8
  nnodes: 2
  val_freq: -1             # origin:5; -1 to disable
  val_before_train: false
  val_only: false
  val_generations_to_log: 1
  save_freq: 50            # origin 5; -1 to disable
  save_limit: 3           # -1 to disable
  save_checkpoint_path: null
  load_checkpoint_path: null`

------------------------------------------------------------------

------------------------------------------------------------------
and my start sh like this:
`#!/bin/bash
#SBATCH --job-name=train_job       # 任务名
#SBATCH --output=output_%j.txt    # 输出日志文件
#SBATCH --error=error_%j.txt      # 错误日志文件
#SBATCH --partition=embodied      # 使用的分区（partition）
#SBATCH --nodes=2                 # 使用的节点数
#SBATCH --gres=gpu:8              # 每个节点的GPU数量
#SBATCH --cpus-per-task=32        # 每个任务的CPU核心数
#SBATCH --mem=512GB               # 每个节点的内存限制
#SBATCH --ntasks-per-node=1       # 每个节点一个任务
#SBATCH --wait-all-nodes=1        # 等待所有节点就绪
#SBATCH --exclude=g2006           # 禁止使用 g2006

echo "============================================="
echo "作业启动时间: $(date)"
echo "作业ID: $SLURM_JOB_ID"
echo "分区: $SLURM_JOB_PARTITION"
echo "总节点数: $SLURM_JOB_NUM_NODES"
echo "分配的节点列表: $SLURM_JOB_NODELIST"
echo "当前工作目录: $(pwd)"
echo "============================================="
echo ""

echo "===== 加载运行环境 ====="
export LD_LIBRARY_PATH=/home/james159/miniconda3/envs/vilasr/lib/python3.10/site-packages/nvidia/nvjitlink/lib:$LD_LIBRARY_PATH

source /home/james159/miniconda3/etc/profile.d/conda.sh
echo "激活conda环境前的环境列表: $(conda env list | grep '^*')"

conda activate vilasr
if [ $? -ne 0 ]; then
    echo "ERROR: 激活vilasr环境失败！" >&2
    exit 1
fi
echo "成功激活conda环境: $(conda info --envs | grep '^*' | awk '{print $1}')"
echo "当前Python路径: $(which python3)"
echo "当前Ray版本: $(ray --version | head -n 1)"
echo "======================="
echo ""

echo "===== 验证脚本路径 ====="
SCRIPT_DIR="/home/james159/hyb/ViLaSR"
START_HEAD="$SCRIPT_DIR/start-head.sh"
START_WORKER="$SCRIPT_DIR/start-worker.sh"

echo "脚本根目录: $SCRIPT_DIR"
echo "头节点脚本路径: $START_HEAD"
echo "工作节点脚本路径: $START_WORKER"

if [ ! -d "$SCRIPT_DIR" ]; then
    echo "ERROR: 脚本根目录 $SCRIPT_DIR 不存在！" >&2
    exit 1
fi
echo "脚本根目录存在: 成功"

if [ ! -f "$START_HEAD" ]; then
    echo "ERROR: 头节点脚本 $START_HEAD 不存在！" >&2
    exit 1
fi
if [ ! -x "$START_HEAD" ]; then
    echo "ERROR: 头节点脚本 $START_HEAD 无执行权限！" >&2
    exit 1
fi
echo "头节点脚本: 存在且可执行"

if [ ! -f "$START_WORKER" ]; then
    echo "ERROR: 工作节点脚本 $START_WORKER 不存在！" >&2
    exit 1
fi
if [ ! -x "$START_WORKER" ]; then
    echo "ERROR: 工作节点脚本 $START_WORKER 无执行权限！" >&2
    exit 1
fi
echo "工作节点脚本: 存在且可执行"
echo "======================="
echo ""

echo "===== 确定头节点信息 ====="
head_node=$(scontrol show hostnames "$SLURM_JOB_NODELIST" | head -n 1)
echo "从节点列表中获取头节点名: $head_node"

head_node_ip=$(srun --nodes=1 --ntasks=1 -w "$head_node" hostname -I | awk '{print $1}')
if [ -z "$head_node_ip" ]; then
    echo "ERROR: 无法获取头节点 $head_node 的IP地址！" >&2
    exit 1
fi
echo "头节点 $head_node 的IP地址: $head_node_ip"
echo "头节点网络信息: $(srun --nodes=1 --ntasks=1 -w "$head_node" ip addr show | grep 'inet ' | awk '{print $2}')"

port=$((10000 + SLURM_JOB_ID % 40000))
if [ $port -lt 10000 ] || [ $port -gt 50000 ]; then
    echo "ERROR: 生成的端口 $port 超出安全范围（10000-50000）！" >&2
    exit 1
fi
echo "生成唯一通信端口: $port"
echo "======================="
echo ""

echo "===== 启动头节点 ====="
echo "启动命令: srun --nodes=1 --ntasks=1 -w $head_node bash -c \"$START_HEAD $head_node_ip $port\""
srun --nodes=1 --ntasks=1 -w "$head_node" \
    bash -c "$START_HEAD $head_node_ip $port" &
head_pid=$!
echo "头节点启动进程ID: $head_pid"
echo "等待头节点初始化（15秒）..."
sleep 15

if ! ps -p $head_pid > /dev/null; then
    echo "ERROR: 头节点进程（PID: $head_pid）已退出！" >&2
    exit 1
fi
echo "头节点进程存活: 成功"
echo "======================="
echo ""

echo "===== 启动工作节点 ====="
worker_nodes=$(scontrol show hostnames "$SLURM_JOB_NODELIST" | tail -n +2)
worker_num=$(echo "$worker_nodes" | wc -l | awk '{print $1}')  # 更可靠的数量计算
echo "工作节点列表: $worker_nodes"
echo "工作节点数量: $worker_num"

if [ $worker_num -gt 0 ]; then
    echo "启动命令: srun --nodes=$worker_num --ntasks=$worker_num -w $worker_nodes bash -c \"$START_WORKER $head_node_ip $port\""
    srun --nodes=$worker_num --ntasks=$worker_num -w "$worker_nodes" \
        bash -c "$START_WORKER $head_node_ip $port" &
    worker_pid=$!
    echo "工作节点启动进程ID: $worker_pid"
    echo "等待工作节点初始化（10秒）..."
    sleep 10

    if ! ps -p $worker_pid > /dev/null; then
        echo "ERROR: 工作节点进程（PID: $worker_pid）已退出！" >&2
        exit 1
    fi
    echo "工作节点进程存活: 成功"
else
    echo "无工作节点需要启动"
fi
echo "======================="
echo ""

echo "===== 执行训练脚本 ====="
echo "切换到工作目录: /home/james159/hyb/ViLaSR"
cd /home/james159/hyb/ViLaSR || exit 1
echo "当前目录: $(pwd)"
echo "训练命令: bash train/rl/train_grpo_multi.sh"
echo "RAY集群地址: $head_node_ip:$port"
export RAY_ADDRESS="$head_node_ip:$port"
echo "已设置环境变量RAY_ADDRESS: $RAY_ADDRESS"

echo "开始训练（$(date)）..."
bash train/rl/train_grpo_multi.sh
train_exit_code=$?
if [ $train_exit_code -ne 0 ]; then
    echo "ERROR: 训练脚本执行失败，退出码: $train_exit_code" >&2
    exit 1
fi
echo "======================="
echo ""

echo "===== 训练完成 ====="
echo "训练结束时间: $(date)"
echo "总作业时长: $(( ($(date +%s) - $SLURM_JOB_START_TIME) / 60 )) 分钟"
echo "清理Ray资源: 执行ray stop（头节点和工作节点）"
srun --nodes=1 --ntasks=1 -w "$head_node" ray stop > /dev/null 2>&1
if [ $worker_num -gt 0 ]; then
    srun --nodes=$worker_num --ntasks=$worker_num -w "$worker_nodes" ray stop > /dev/null 2>&1
fi
echo "Training completed successfully!"`

------------------------------------------------------------------
